Add OpInfo and ref for linspace and logspace

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #82284
* __->__ #81826
* #81734

Implements linspace with arange, and logspace with linspace.
- Implements a more precise path in linspace's ref when dtype is integral to avoid off-by-one issues when output of computation is casted to int. The trade off is that there's an increased chance of overflow.
- Files several issues #82242, #82230, #81996, on preexisting issues with the linspace and logspace. These mainly concern when dtype is integral - the affect tests are xfailed in this PR.
- Fixes the check that the reference implementation is closer to precise implementation than torch implementation to also update the dtype kwarg to the precise dtype.

TODO:
- ~support negative bases~ (not in this PR)
- ~support complex. Since arange does not support complex, but linspace does, one solution is to just call linspace separately on the real and imag components and sum the results in the end~ (not in this PR)
- ~default dtypes need to be explicitly handled since computation is done in a different dtype than result~ (done)

soulitzer committed:
Add OpInfo and ref for linspace

[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace"




[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace"




[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"




[ghstack-poisoned]

soulitzer commented on torch/_refs/__init__.py:
```diff
@@ -3055,6 +3057,88 @@ def arange(
+    pin_memory: bool = False,
+    requires_grad: bool = False,
+) -> TensorLikeType:
+    assert not isinstance(start, complex)
```
TODO: replace with
```python
if any(isinstance(a, complex) for a in (start, end, steps)):
    raise NotImplementedError
if steps < 0:
    raise NotImplementedError
```

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"


Implements linspace with arange, and logspace with linspace.

TODO:
- ~support negative bases~ (not in this PR)
- ~support complex. Since arange does not support complex, but linspace does, one solution is to just call linspace separately on the real and imag components and sum the results in the end~ (not in this PR)
- ~default dtypes need to be explicitly handled since computation is done in a different dtype than result~ (done)

[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"


Implements linspace with arange, and logspace with linspace.

TODO:
- ~support negative bases~ (not in this PR)
- ~support complex. Since arange does not support complex, but linspace does, one solution is to just call linspace separately on the real and imag components and sum the results in the end~ (not in this PR)
- ~default dtypes need to be explicitly handled since computation is done in a different dtype than result~ (done)

[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"


Implements linspace with arange, and logspace with linspace.

TODO:
- ~support negative bases~ (not in this PR)
- ~support complex. Since arange does not support complex, but linspace does, one solution is to just call linspace separately on the real and imag components and sum the results in the end~ (not in this PR)
- ~default dtypes need to be explicitly handled since computation is done in a different dtype than result~ (done)

[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"


Implements linspace with arange, and logspace with linspace.

TODO:
- ~support negative bases~ (not in this PR)
- ~support complex. Since arange does not support complex, but linspace does, one solution is to just call linspace separately on the real and imag components and sum the results in the end~ (not in this PR)
- ~default dtypes need to be explicitly handled since computation is done in a different dtype than result~ (done)

[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"


Implements linspace with arange, and logspace with linspace.

TODO:
- ~support negative bases~ (not in this PR)
- ~support complex. Since arange does not support complex, but linspace does, one solution is to just call linspace separately on the real and imag components and sum the results in the end~ (not in this PR)
- ~default dtypes need to be explicitly handled since computation is done in a different dtype than result~ (done)

[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"


Implements linspace with arange, and logspace with linspace.

TODO:
- ~support negative bases~ (not in this PR)
- ~support complex. Since arange does not support complex, but linspace does, one solution is to just call linspace separately on the real and imag components and sum the results in the end~ (not in this PR)
- ~default dtypes need to be explicitly handled since computation is done in a different dtype than result~ (done)

[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"


Implements linspace with arange, and logspace with linspace.
- Implements a more precise path in linspace's ref when dtype is integral to avoid off-by-one issues when output of computation is casted to int. The trade off is that there's an increased chance of overflow.
- Files several issues #82242, #82230, #81996, on preexisting issues with the linspace and logspace. These mainly concern when dtype is integral - the affect tests are xfailed in this PR.
- Fixes the check that the reference implementation is closer to precise implementation than torch implementation to also update the dtype kwarg to the precise dtype.

TODO:
- ~support negative bases~ (not in this PR)
- ~support complex. Since arange does not support complex, but linspace does, one solution is to just call linspace separately on the real and imag components and sum the results in the end~ (not in this PR)
- ~default dtypes need to be explicitly handled since computation is done in a different dtype than result~ (done)

[ghstack-poisoned]

soulitzer commented:
Ready for review, @ngimel!

soulitzer commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -20577,6 +20679,78 @@ def __init__(
+        decorators=(
+            DecorateInfo(
+                toleranceOverride({
+                    torch.float16: tol(atol=1e-02, rtol=1e-02),
```
It may be possible to remove this, will need to check

soulitzer replied:
removed.

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -1644,6 +1646,10 @@ def all(
     return result
 
 
+# Saves Python all
```
comment is wrong?

ngimel reviewed (approved):
This looks good, but oh my, so many skips!

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"


Implements linspace with arange, and logspace with linspace.
- Implements a more precise path in linspace's ref when dtype is integral to avoid off-by-one issues when output of computation is casted to int. The trade off is that there's an increased chance of overflow.
- Files several issues #82242, #82230, #81996, on preexisting issues with the linspace and logspace. These mainly concern when dtype is integral - the affect tests are xfailed in this PR.
- Fixes the check that the reference implementation is closer to precise implementation than torch implementation to also update the dtype kwarg to the precise dtype.

TODO:
- ~support negative bases~ (not in this PR)
- ~support complex. Since arange does not support complex, but linspace does, one solution is to just call linspace separately on the real and imag components and sum the results in the end~ (not in this PR)
- ~default dtypes need to be explicitly handled since computation is done in a different dtype than result~ (done)

[ghstack-poisoned]

soulitzer committed:
Update on "Add OpInfo and ref for linspace and logspace"


Implements linspace with arange, and logspace with linspace.
- Implements a more precise path in linspace's ref when dtype is integral to avoid off-by-one issues when output of computation is casted to int. The trade off is that there's an increased chance of overflow.
- Files several issues #82242, #82230, #81996, on preexisting issues with the linspace and logspace. These mainly concern when dtype is integral - the affect tests are xfailed in this PR.
- Fixes the check that the reference implementation is closer to precise implementation than torch implementation to also update the dtype kwarg to the precise dtype.

TODO:
- ~support negative bases~ (not in this PR)
- ~support complex. Since arange does not support complex, but linspace does, one solution is to just call linspace separately on the real and imag components and sum the results in the end~ (not in this PR)
- ~default dtypes need to be explicitly handled since computation is done in a different dtype than result~ (done)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g
